firmware/controllers/gauges/tachometer.cpp

```diff
@@ -12,44 +12,97 @@
 
 #include "tachometer.h"
 #include "trigger_central.h"
-
-#if !EFI_UNIT_TEST
+#include "pwm_generator.h"
 
 EXTERN_ENGINE;
 
-static scheduling_s tachTurnSignalOff;
 
-static void turnTachPinLow(void *) {
-	enginePins.tachOut.setLow();
-}
+static SimplePwm tachControl("tach");
+static bool isTachEnabled;
+
+static void applyTachoPinState(int stateIndex, PwmConfig *state) /* pwm_gen_callback */ {
+	efiAssertVoid(CUSTOM_ERR_6643, stateIndex < PWM_PHASE_MAX_COUNT, "invalid stateIndex");
+	efiAssertVoid(CUSTOM_ERR_PWM_CALLBACK_ASSERT, state->multiChannelStateSequence.waveCount == 1, "invalid idle waveCount");
+	OutputPin *output = state->outputPins[0];
+	int value = state->multiChannelStateSequence.getChannelState(/*channelIndex*/0, stateIndex);
+
+	if (!value || isTachEnabled)
+		output->setValue(value);
+}        
 
 static void tachSignalCallback(trigger_event_e ckpSignalType,
 		uint32_t index, efitick_t edgeTimestamp DECLARE_ENGINE_PARAMETER_SUFFIX) {
-	UNUSED(ckpSignalType);
+	// only process at index configured to avoid too much cpu time for index 0?
 	if (index != (uint32_t)engineConfiguration->tachPulseTriggerIndex) {
 		return;
 	}
-	enginePins.tachOut.setHigh();
-	float durationMs;
+
+#if EFI_UNIT_TEST
+	printf("tachSignalCallback(%d %d)\n", ckpSignalType, index);
+#else
+	UNUSED(ckpSignalType);
+	UNUSED(edgeTimestamp);
+#endif
+
+	// How many tach pulse periods do we have?
+	int periods = engineConfiguration->tachPulsePerRev;
+
+	if(periods == 0){
+		warning(CUSTOM_ERR_6709,"Check Tachometer Pulse per Rev!");
+		return;
+	}
+
+	// What is the angle per tach output period?
+	float duty;
+	float cycleTimeMs = 60000.0 / GET_RPM();
+	float periodTimeMs = cycleTimeMs / periods;
+	float tachFreq = 1000.0 / periodTimeMs;
+	
 	if (engineConfiguration->tachPulseDurationAsDutyCycle) {
-		// todo: implement tachPulseDurationAsDutyCycle
-		durationMs = engineConfiguration->tachPulseDuractionMs;
+		// Simple case - duty explicitly set
+		duty = engineConfiguration->tachPulseDuractionMs;
 	} else {
-		durationMs = engineConfiguration->tachPulseDuractionMs;
+		// Constant high-time mode - compute the correct duty cycle
+		duty = engineConfiguration->tachPulseDuractionMs / periodTimeMs;
 	}
-	engine->executor.scheduleForLater(&tachTurnSignalOff, (int)MS2US(durationMs), &turnTachPinLow);
+
+	// limit to 10..90% duty
```

Comment:
why limit duty? what is special about 10%? Please add code comments explaining the intention